scalyr_agent/date_parsing_utils.py

```diff
@@ -0,0 +1,465 @@
+            int(result2[2]),
+        )
+    except Exception:
+        return None
+
+    # now add the fractional part
+    if len(parts) > 1:
+        fractions = parts[1]
+        # if we had a fractional component it should terminate in a Z
+        if not fractions.endswith("Z"):
+            # we don't handle non UTC timezones yet
+            if any(c in fractions for c in "+-"):
+                return None
+            return dt
+
+        # remove the Z and just process the fraction.
+        fractions = fractions[:-1]
+        to_micros = 6 - len(fractions)
+        micro = int(int(fractions) * 10 ** to_micros)
+        dt = dt.replace(microsecond=micro)
+
+    return dt
+
+
+def _rfc3339_to_datetime_udatetime(string):
+    """Returns a date time from a rfc3339 formatted timestamp.
+    We have to do some tricksy things to support python 2.4, which doesn't support
+    datetime.strptime or the fractional component %f in format strings
+    This doesn't do any complex testing and assumes the string is well formed
+    and in UTC (e.g. uses Z at the end rather than a time offset)
+    @param string: a date/time in rfc3339 format, e.g. 2015-08-03T09:12:43.143757463Z
+    @rtype datetime.datetime
+    """
+    # split the string in to main time and fractional component
+    parts = string.split(".")
+
+    # it's possible that the time does not have a fractional component
+    # e.g 2015-08-03T09:12:43Z, in this case 'parts' will only have a
+    # single element that should end in Z.  Strip the Z if it exists
+    # so we can use the same format string for processing the main
+    # date+time regardless of whether the time has a fractional component.
+    if parts[0].endswith("Z"):
+        parts[0] = parts[0][:-1]
+
+    # create a datetime object
+    try:
+        dt = udatetime.from_string(parts[0])
+        # NOTE: At this point we don't support timezones
+        dt = dt.replace(tzinfo=None)
+    except ValueError:
+        return None
+
+    # now add the fractional part
+    if len(parts) > 1:
+        fractions = parts[1]
+        # if we had a fractional component it should terminate in a Z
+        if not fractions.endswith("Z"):
+            # we don't handle non UTC timezones yet
+            if any(c in fractions for c in "+-"):
+                return None
+            return dt
+
+        # remove the Z and just process the fraction.
+        fractions = fractions[:-1]
+        to_micros = 6 - len(fractions)
+        micro = int(int(fractions) * 10 ** to_micros)
+        # NOTE(Tomaz): dt.replace is quite slow...
+        dt = dt.replace(microsecond=micro)
+
+    return dt
+
+
+if udatetime:
+    rfc3339_to_nanoseconds_since_epoch = _rfc3339_to_nanoseconds_since_epoch_udatetime
+    rfc3339_to_datetime = _rfc3339_to_datetime_udatetime
+else:
+    rfc3339_to_nanoseconds_since_epoch = (
+        _rfc3339_to_nanoseconds_since_epoch_string_split
+    )
+    rfc3339_to_datetime = _rfc3339_to_datetime_strptime
```

Comment:
Is there a reason we are using `_rfc3339_to_datetime_strptime` instead of `_rfc3339_to_datetime_string_split` here?

Reply:
That was a typo.

Resolved in 2f7970c1a9a92e1ff37c74d8ac09033021884c75.